src/core/config.cpp

```diff
@@ -126,14 +126,30 @@ bool Config::sip003() {
 }
 
 string Config::SHA224(const string &message) {
-    uint8_t digest[SHA224_DIGEST_LENGTH];
-    SHA256_CTX ctx;
-    SHA224_Init(&ctx);
-    SHA224_Update(&ctx, message.c_str(), message.length());
-    SHA224_Final(digest, &ctx);
-    char mdString[(SHA224_DIGEST_LENGTH << 1) + 1];
-    for (int i = 0; i < SHA224_DIGEST_LENGTH; ++i) {
+    uint8_t digest[EVP_MAX_MD_SIZE];
+    unsigned int digest_len;
+    EVP_MD_CTX *ctx;
+    if ((ctx = EVP_MD_CTX_new()) == NULL) {
+        EVP_MD_CTX_free(ctx);
+        throw runtime_error("Could not create hash context");
+    }
+    if (!EVP_DigestInit_ex(ctx, EVP_sha224(), NULL)) {
+        EVP_MD_CTX_free(ctx);
+        throw runtime_error("Could not initialize hash context");
+    }
+    if (!EVP_DigestUpdate(ctx, message.c_str(), message.length())) {
+        EVP_MD_CTX_free(ctx);
+        throw runtime_error("Could not update hash");
+    }
+    if (!EVP_DigestFinal_ex(ctx, digest, &digest_len)) {
+        EVP_MD_CTX_free(ctx);
+        throw runtime_error("Could not output hash");
+    }
+
+    char mdString[(digest_len << 1) + 1];
```

Comment:
MSVC (cl.exe) doesn't like variable length arrays because it's a C99 feature (not in C++). That's why the CI failed with [`D:\a\1\s\src\core\config.cpp(149,37): error C2131: expression did not evaluate to a constant [D:\a\1\s\trojan.vcxproj]
`](https://dev.azure.com/GreaterFire/Trojan-GFW/_build/results?buildId=147&view=logs&j=0bc77094-9fcd-5c38-f6e4-27d2ae131589&t=0a8c9221-3327-51b2-21a2-2509bc3fa943&l=77).